datafusion/src/physical_plan/planner.rs

```diff
@@ -296,520 +301,536 @@ impl DefaultPhysicalPlanner {
     }
 
     /// Create a physical plan from a logical plan
-    fn create_initial_plan(
-        &self,
-        logical_plan: &LogicalPlan,
-        ctx_state: &ExecutionContextState,
-    ) -> Result<Arc<dyn ExecutionPlan>> {
-        let batch_size = ctx_state.config.batch_size;
-
-        match logical_plan {
-            LogicalPlan::TableScan {
-                source,
-                projection,
-                filters,
-                limit,
-                ..
-            } => {
-                // Remove all qualifiers from the scan as the provider
-                // doesn't know (nor should care) how the relation was
-                // referred to in the query
-                let filters = unnormalize_cols(filters.iter().cloned());
-                source.scan(projection, batch_size, &filters, *limit)
-            }
-            LogicalPlan::Window {
-                input, window_expr, ..
-            } => {
-                if window_expr.is_empty() {
-                    return Err(DataFusionError::Internal(
-                        "Impossibly got empty window expression".to_owned(),
-                    ));
+    fn create_initial_plan<'a>(
+        &'a self,
+        logical_plan: &'a LogicalPlan,
+        ctx_state: &'a ExecutionContextState,
+    ) -> BoxFuture<'a, Result<Arc<dyn ExecutionPlan>>> {
```

Comment:
This shows as a big change but only because the function body needs to be wrapped into a `BoxFuture` with `async { .. }.boxed()`

Reply:
why can't this function signature be made `async fn create_initial_plan....` ? Why does it need a boxed future?

FWIW I found it easier to understand the diffs without whitespace: https://github.com/apache/arrow-datafusion/pull/1013/files?w=1

Reply:
Because of async's internals: https://rust-lang.github.io/async-book/07_workarounds/04_recursion.html

Reply:
Because of async's internals: https://rust-lang.github.io/async-book/07_workarounds/04_recursion.html

Reply:
Ah, I see -- so that it can recursively call itself. 👍